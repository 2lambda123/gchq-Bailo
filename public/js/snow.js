const TOTAL_SNOWFLAKES = 200

let snowDiv = document.getElementById('snowContainer')

function embRand(a, b) {
  return Math.floor(Math.random() * (b - a + 1)) + a
}

if (!snowDiv) {
  const snowflakes = new Array(TOTAL_SNOWFLAKES).fill(`<i class="snow"></i>`).join('\n')

  const snow = `
    .snow {
      position: absolute;
      width: 10px;
      height: 10px;
      background: white;
      border-radius: 50%;
      margin-top:-10px;
    }

    #snowContainer {
      position: fixed;
      left: 0;
      top: 0;
      bottom: 0;
      width: 100vw;
      height: 100vh;
      overflow: hidden;
      z-index: 9999999;
      pointer-events: none;
    }
  `

  const transitions = new Array(TOTAL_SNOWFLAKES)
    .fill(undefined)
    .map((_, i) => {
      const rndX = embRand(0, 1000000) * 0.0001
      const rndO = embRand(-100000, 100000) * 0.0001
      const rndT = (embRand(3, 8) * 10).toFixed(2)
      const rndS = (embRand(0, 10000) * 0.0001).toFixed(2)

      return `
      .snow:nth-child(${i}) {
        opacity: ${(embRand(1, 10000) * 0.0001).toFixed(2)};
        transform: translate(${rndX.toFixed(2)}vw,-10px) scale(${rndS});
        animation: fall-${i} ${embRand(10, 30)}s -${embRand(0, 30)}s linear infinite;
      }

      @keyframes fall-${i}{
        ${rndT}%{
          transform: translate(${(rndX + rndO).toFixed(2)}vw,${rndT}vh) scale(${rndS});
        } to {
          transform: translate(${(rndX + rndO / 2).toFixed(2)}vw, 105vh) scale(${rndS});
        }
      }
    `
    })
    .join('\n')

  snowDiv = document.createElement('div')
  snowDiv.id = 'snowContainer'

  snowDiv.innerHTML = `
    <style>
      ${snow}
      ${transitions}
    </style>
    ${snowflakes}
  `
  document.body.appendChild(snowDiv)
}
