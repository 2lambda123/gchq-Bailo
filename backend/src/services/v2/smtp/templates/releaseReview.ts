import dedent from 'dedent-js'
import mjml2html from 'mjml'

import { ReviewKind } from '../../../../types/v2/enums.js'
import { BaseEmailTemplate } from './baseEmailTemplate.js'

export class ReleaseReviewEmail extends BaseEmailTemplate {
  setTo(emailAddress: string) {
    this.to = emailAddress
  }

  setSubject(resourceName: string, reviewerRole: string) {
    this.subject = dedent(`
    ${reviewerRole.toUpperCase()}: You have been requested to review '${resourceName}' on Bailo
  `)
  }

  setText(releaseName: string, modelId: string, baseUrl: string, author: string) {
    // V2 change- we don't store the author of a release
    // TODO - Replace with URL to specific model release
    this.text = dedent(`
    You have been requested to review '${releaseName}' on Bailo.

    Review Category: '${ReviewKind.Release}'
    Author: '${author}'

    Open ${ReviewKind.Release}: ${baseUrl}/model/${modelId}
    See Reviews: ${baseUrl}/review
  `)
  }

  setHtml(releaseName: string, modelId: string, baseUrl: string, author: string) {
    this.html = mjml2html(`
    <mjml>
      <mj-head>
        <mj-style>
          .diagonal {
          top: 0;
          bottom: 0;
          right: 0;
          left: 0;
          width: 100%;
          height: 50px;
          background-color: #f0f0f0;
          z-index: 0;
          transform: skewY(2deg);
          transform-origin: top right;
          height: 12px;
          }
          .gradient-bg {
          background: linear-gradient(276deg, rgba(214,37,96) 0%, rgba(84,39,142) 100%);
          }
          .gradient-bg-border {
          background: linear-gradient(276deg, rgba(209,73,118) 0%, rgba(126,66,204) 100%);
          height: 12px;
          }
          .text {
          height: 12px;
          }
        </mj-style>
      </mj-head>
      <mj-body>
        <mj-wrapper background-color="#9e60f2" padding-bottom="0px" padding-top="0px">
          <mj-section padding-bottom="5px" css-class='gradient-bg'>
            <mj-column width="100%">
              <mj-text css-class='text' align="center" font-weight="bold" color="#FFF" font-size="25px" font-family="Helvetica" padding-left="25px" padding-right="25px" padding-bottom="10px" padding-top="10px">Bailo
              </mj-text>
            </mj-column>
          </mj-section>
          <mj-section padding-bottom="5px" css-class='gradient-bg'>
            <mj-column width="100%">
              <mj-text align="center" color="#FFF" font-size="20px" font-family="Helvetica" padding-left="25px" padding-right="25px" padding-bottom="30px" padding-top="10px">You have been requested to review a <span style="font-weight:bold">${ReviewKind.Release.toLowerCase()}.</span>
              </mj-text>
            </mj-column>
          </mj-section>
          <mj-section css-class='diagonal gradient-bg'>
          </mj-section>
          <mj-section css-class='diagonal gradient-bg-border'>
          </mj-section>
          <mj-section background-color="#9e60f2" padding-bottom="15px">
            <mj-column>
              <mj-text align="center" color="#FFF" font-size="15px" font-family="Ubuntu, Helvetica, Arial, sans-serif" padding-left="25px" padding-right="25px" padding-bottom="0px"><strong>Model Name</strong></mj-text>
              <mj-text align="center" color="#FFF" font-size="13px" font-family="Helvetica" padding-left="25px" padding-right="25px" padding-bottom="20px" padding-top="10px">${releaseName}</mj-text>
            </mj-column>
            <mj-column>
              <mj-text align="center" color="#FFF" font-size="15px" font-family="Ubuntu, Helvetica, Arial, sans-serif" padding-left="25px" padding-right="25px" padding-bottom="0px"><strong>Review Category</strong></mj-text>
              <mj-text align="center" color="#FFF" font-size="13px" font-family="Helvetica" padding-left="25px" padding-right="25px" padding-bottom="20px" padding-top="10px">${
                ReviewKind.Release
              }</mj-text>
            </mj-column>
            <mj-column>
              <mj-text align="center" color="#FFF" font-size="15px" font-family="Ubuntu, Helvetica, Arial, sans-serif" padding-left="25px" padding-right="25px" padding-bottom="0px"><strong>Author</strong></mj-text>
              <mj-text align="center" color="#FFF" font-size="13px" font-family="Helvetica" padding-left="25px" padding-right="25px" padding-bottom="20px" padding-top="10px">${author}</mj-text>
            </mj-column>
          </mj-section>
          <mj-section background-color="#54278e" padding-bottom="20px" padding-top="20px">
            <mj-column width="50%">
              <mj-button background-color="#54278e" border="solid 2px white" color="#FFF" font-size="14px" align="center" font-weight="bold" padding="15px 30px" border-radius="5px" href="${baseUrl}/model/${modelId}" font-family="Helvetica" padding-left="25px" padding-right="25px" padding-bottom="10px">Open ${
                ReviewKind.Release
              }</mj-button>
            </mj-column>
            <mj-column width="50%">
              <mj-button background-color="#54278e" border="solid 2px white" color="#FFF" font-size="14px" align="center" font-weight="bold" padding="15px 30px" border-radius="5px" href="${baseUrl}/review" font-family="Helvetica" padding-left="25px" padding-right="25px" padding-bottom="12px">See Reviews</mj-button>
            </mj-column>
          </mj-section>
        </mj-wrapper>
      </mj-body>
    </mjml>
  `).html
  }
}
