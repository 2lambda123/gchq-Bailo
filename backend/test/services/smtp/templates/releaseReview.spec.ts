import { describe, expect, test, vi } from 'vitest'

import { ReleaseReviewEmail } from '../../../../src/services/v2/smtp/templates/releaseReview.js'

vi.mock('../../../../src/utils/v2/config.js', () => {
  return {
    __esModule: true,
    default: {
      smtp: {
        from: '"Bailo 📝" <bailo@example.org>',
      },
    },
  }
})

const mockDedent = vi.hoisted(() => {
  const model: any = vi.fn()

  return model
})
vi.mock('dedent-js', async () => ({
  default: mockDedent,
}))

const mockMjml = vi.hoisted(() => {
  const model: any = vi.fn(() => ({
    html: 'html',
  }))

  return model
})
vi.mock('mjml', async () => ({
  default: mockMjml,
}))

describe('services > smtp > templates > releaseReview', () => {
  test('that the subject is set', async () => {
    const testEmail = new ReleaseReviewEmail()
    testEmail.setSubject('resource', 'role')
    expect(mockDedent.mock.calls).toMatchSnapshot()
  })

  test('that the html is set', async () => {
    const testEmail = new ReleaseReviewEmail()
    testEmail.setHtml('releaseName', 'modelId', '/base/url', 'author')
    expect(mockDedent.mock.calls).toMatchSnapshot()
  })

  test('that the text is set', async () => {
    const testEmail = new ReleaseReviewEmail()
    testEmail.setText('releaseName', 'modelId', '/base/url', 'author')
    expect(mockDedent.mock.calls).toMatchSnapshot()
  })

  test('that the html is set', async () => {
    const testEmail = new ReleaseReviewEmail()
    testEmail.setHtml('releaseName', 'modelId', '/base/url', 'author')
    expect(mockDedent.mock.calls).toMatchSnapshot()
  })

  test('that the to field is set', async () => {
    const testEmail = new ReleaseReviewEmail()
    testEmail.setTo('email@address')
    expect(testEmail.to).toBe('email@address')
  })
})
