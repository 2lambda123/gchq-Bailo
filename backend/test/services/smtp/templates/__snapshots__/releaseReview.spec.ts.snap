// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > smtp > templates > releaseReview > that the html is set 1`] = `
[
  [
    "
    <mjml>
      <mj-body background-color=\\"#ccd3e0\\">
        <mj-section background-color=\\"#ccd3e0\\" padding-bottom=\\"20px\\" padding-top=\\"20px\\">
        </mj-section>
        
    <mj-section background-color=\\"#27598e\\" padding-bottom=\\"5px\\" padding-top=\\"20px\\">
      <mj-column width=\\"100%\\">
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"28px\\" padding-top=\\"28px\\"><span style=\\"font-size:20px; font-weight:bold\\">You have been requested to review a release.</span>
        </mj-text>
      </mj-column>
    </mj-section>
    <mj-section background-color=\\"#356cc7\\" padding-bottom=\\"15px\\">
      <mj-column>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"15px\\" font-family=\\"Ubuntu, Helvetica, Arial, sans-serif\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"0px\\"><strong>Model Name</strong></mj-text>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"20px\\" padding-top=\\"10px\\">releaseName</mj-text>
      </mj-column>
      <mj-column>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"15px\\" font-family=\\"Ubuntu, Helvetica, Arial, sans-serif\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"0px\\"><strong>Review Category</strong></mj-text>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"20px\\" padding-top=\\"10px\\">release</mj-text>
      </mj-column>
      <mj-column>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"15px\\" font-family=\\"Ubuntu, Helvetica, Arial, sans-serif\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"0px\\"><strong>Author</strong></mj-text>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"20px\\" padding-top=\\"10px\\">author</mj-text>
      </mj-column>
    </mj-section>
    <mj-section background-color=\\"#27598e\\" padding-bottom=\\"20px\\" padding-top=\\"20px\\">
      <mj-column width=\\"50%\\">
        <mj-button background-color=\\"#f37f58\\" color=\\"#FFF\\" font-size=\\"14px\\" align=\\"center\\" font-weight=\\"bold\\" border=\\"none\\" padding=\\"15px 30px\\" border-radius=\\"10px\\" href=\\"/base/url/model/modelId\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"10px\\">Open release</mj-button>
      </mj-column>
      <mj-column width=\\"50%\\">
        <mj-button background-color=\\"#f37f58\\" color=\\"#FFF\\" font-size=\\"14px\\" align=\\"center\\" font-weight=\\"bold\\" border=\\"none\\" padding=\\"15px 30px\\" border-radius=\\"10px\\" href=\\"/base/url/review\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"12px\\">See Reviews</mj-button>
      </mj-column>
    </mj-section>
  
      </mj-body>
    </mjml>
  ",
  ],
]
`;

exports[`services > smtp > templates > releaseReview > that the html is set 2`] = `
[
  [
    "
    <mjml>
      <mj-body background-color=\\"#ccd3e0\\">
        <mj-section background-color=\\"#ccd3e0\\" padding-bottom=\\"20px\\" padding-top=\\"20px\\">
        </mj-section>
        
    <mj-section background-color=\\"#27598e\\" padding-bottom=\\"5px\\" padding-top=\\"20px\\">
      <mj-column width=\\"100%\\">
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"28px\\" padding-top=\\"28px\\"><span style=\\"font-size:20px; font-weight:bold\\">You have been requested to review a release.</span>
        </mj-text>
      </mj-column>
    </mj-section>
    <mj-section background-color=\\"#356cc7\\" padding-bottom=\\"15px\\">
      <mj-column>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"15px\\" font-family=\\"Ubuntu, Helvetica, Arial, sans-serif\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"0px\\"><strong>Model Name</strong></mj-text>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"20px\\" padding-top=\\"10px\\">releaseName</mj-text>
      </mj-column>
      <mj-column>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"15px\\" font-family=\\"Ubuntu, Helvetica, Arial, sans-serif\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"0px\\"><strong>Review Category</strong></mj-text>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"20px\\" padding-top=\\"10px\\">release</mj-text>
      </mj-column>
      <mj-column>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"15px\\" font-family=\\"Ubuntu, Helvetica, Arial, sans-serif\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"0px\\"><strong>Author</strong></mj-text>
        <mj-text align=\\"center\\" color=\\"#FFF\\" font-size=\\"13px\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"20px\\" padding-top=\\"10px\\">author</mj-text>
      </mj-column>
    </mj-section>
    <mj-section background-color=\\"#27598e\\" padding-bottom=\\"20px\\" padding-top=\\"20px\\">
      <mj-column width=\\"50%\\">
        <mj-button background-color=\\"#f37f58\\" color=\\"#FFF\\" font-size=\\"14px\\" align=\\"center\\" font-weight=\\"bold\\" border=\\"none\\" padding=\\"15px 30px\\" border-radius=\\"10px\\" href=\\"/base/url/model/modelId\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"10px\\">Open release</mj-button>
      </mj-column>
      <mj-column width=\\"50%\\">
        <mj-button background-color=\\"#f37f58\\" color=\\"#FFF\\" font-size=\\"14px\\" align=\\"center\\" font-weight=\\"bold\\" border=\\"none\\" padding=\\"15px 30px\\" border-radius=\\"10px\\" href=\\"/base/url/review\\" font-family=\\"Helvetica\\" padding-left=\\"25px\\" padding-right=\\"25px\\" padding-bottom=\\"12px\\">See Reviews</mj-button>
      </mj-column>
    </mj-section>
  
      </mj-body>
    </mjml>
  ",
  ],
]
`;

exports[`services > smtp > templates > releaseReview > that the subject is set 1`] = `
[
  [
    "
    ROLE: You have been requested to review 'resource' on Bailo
  ",
  ],
]
`;

exports[`services > smtp > templates > releaseReview > that the text is set 1`] = `
[
  [
    "
    You have been requested to review 'releaseName' on Bailo.

    Review Category: 'release'
    Author: 'author'

    Open release: /base/url/model/modelId
    See Reviews: /base/url/review
  ",
  ],
]
`;
